util/dvsim/LsfLauncher.py

```diff
@@ -0,0 +1,379 @@
+        job_name = self.deploy.job_name
+        cfg = self.deploy.sim_cfg
+
+        if job_name not in LsfLauncher.jobs[cfg]:
+            LsfLauncher.jobs[cfg][job_name] = []
+        LsfLauncher.jobs[cfg][job_name].append(self)
+
+        # The actual launching of the bsub command cannot happen until the
+        # Scheduler has dispatched ALL jobs in the array.
+        if len(LsfLauncher.jobs[cfg][job_name]) < self.deploy.job_total:
+            return
+
+        # Write the job array script.
+        job_script_wo_idx = Path(LsfLauncher.jobs_dir[cfg], job_name)
+        try:
+            with open(job_script_wo_idx, "w", encoding='utf-8') as f:
+                f.writelines(self.make_job_array_script_text(cfg, job_name))
+        except IOError as e:
+            err_msg = "ERROR: Failed to write job script {}:\n{}".format(
+                job_script_wo_idx, e)
+            self._kill_job_array(err_msg)
+            raise LauncherError(err_msg)
+
+        # Update the shell's env vars with self.exports. Values in exports must
+        # replace the values in the shell's env vars if the keys match.
+        exports = os.environ.copy()
+        if self.deploy.exports:
+            exports.update(self.deploy.exports)
+
+        # Clear the magic MAKEFLAGS variable from exports if necessary. This
+        # variable is used by recursive Make calls to pass variables from one
+        # level to the next. Here, self.cmd is a call to Make but it's
+        # logically a top-level invocation: we don't want to pollute the flow's
+        # Makefile with Make variables from any wrapper that called dvsim.
+        if 'MAKEFLAGS' in exports:
+            del exports['MAKEFLAGS']
+
+        self._dump_env_vars(exports)
+
+        # TODO: Arbitrarily set the max slot-limit to 100.
+        job_array = "{}[1-{}]".format(job_name, self.deploy.job_total)
+        if self.deploy.job_total > 100:
+            job_array += "%100"
+
+        # TODO: This needs to be moved to a HJson.
+        if self.deploy.sim_cfg.tool == "vcs":
+            job_rusage = "\'rusage[vcssim=1,vcssim_dynamic=1:duration=1]\'"
+
+        elif self.deploy.sim_cfg.tool == "xcelium":
+            job_rusage = "\'rusage[xcelium=1,xcelium_dynamic=1:duration=1]\'"
+
+        else:
+            job_rusage = None
+
+        # Launch the job array.
+        cmd = [
+            "bsub",
+            # TODO: LSF project name could be site specific!
+            "-P",
+            cfg.project,
+            "-J",
+            "\"{}\"".format(job_array),
+            "-oo",
+            "\"{}.%I.out\"".format(job_script_wo_idx),
+            "-eo",
+            "\"{}.%I.out\"".format(job_script_wo_idx)
+        ]
+
+        if job_rusage:
+            cmd.extend(["-R", job_rusage])
+
+        cmd.append(
+            "\"/usr/bin/bash {} \\$LSB_JOBINDEX\"".format(job_script_wo_idx))
+
+        try:
+            cmd = " ".join(cmd)
+            p = subprocess.run(cmd,
```

Comment:
Or even better: just remove it and pass cmd (the list) to subprocess.run(), which will then take care of the quoting and other things reliably.

Reply:
Hmm, that's really weird! I would have expected the two things to be equivalent. It might be worth understanding what's going on - maybe try running both types under strace to see what's actually getting exec'd?

Reply:
No need to discuss that forever, but if you're curious: This is exactly one of the cases where passing arguments as lists comes handy and just removing the `cmd = " ".join(cmd)` part should do the trick.

* The command that is ultimately executed by bsub must likely be a single argument, i.e. a single element within the cmd array (`"/usr/bin/bash {} $LSB_JOBINDEX".format(job_script_wo_idx)))`) (This depends a bit on how bsub parses its command line; it could also support or even prefer a list of arguments there.)
* The arguments passed to bsub that modify its own behavior need to be individual arguments, i.e. individual entries within the list.

Ultimately, each list element will be an element in the `argv` array received by bsub.